python/pyarrow/tests/test_csv.py

```diff
@@ -59,6 +59,13 @@ def make_random_csv(num_cols=2, num_rows=10, linesep=u'\r\n'):
     return csv, expected
 
 
+def test_nonlocal_file_errors_nicely():
+    with pytest.raises(pa.lib.ArrowIOError, match="Try using HadoopFileSystem"):
+        read_csv("hdfs://does-not-exist.csv")
+    with pytest.raises(pa.lib.ArrowIOError, match="Try using S3FSWrapper"):
+        read_csv("s3://does-not-exist.csv")
```

Comment:
Neither of these classes are really public APIs, `HadoopFileSystem` is obtained through `pa.hdfs.connect` and `S3FSWrapper` is an implementation detail when the user passes a s3fs file handle. 

A better solution might be to parse the URI and see if there is a "scheme" at all and say that `URI scheme $scheme cannot be accessed on the local filesystem`

Reply:
@pitrou thoughts?

Reply:
Yes, I think rejecting URIs upfront would be better than this.

Reply:
Particularly given https://github.com/apache/arrow/pull/4225, might the right place to reject up front be down in the C++ layer?

Reply:
That's an interesting idea at least.